src/Workspaces/Core/Portable/EmbeddedLanguages/VirtualChars/VirtualCharSequence.cs

```diff
@@ -0,0 +1,147 @@
+        /// </summary>
+        private readonly TextSpan _span;
+
+        private VirtualCharSequence(Chunk sequence) : this(sequence, new TextSpan(0, sequence.Length))
+        {
+        }
+
+        private VirtualCharSequence(Chunk sequence, TextSpan span)
+        {
+            if (span.Start > sequence.Length)
+            {
+                throw new ArgumentException();
+            }
+
+            if (span.End > sequence.Length)
+            {
+                throw new ArgumentException();
+            }
+
+            _leafCharacters = sequence;
+            _span = span;
+        }
+
+        public int Length => _span.Length;
+        public VirtualChar this[int index] => _leafCharacters[_span.Start + index];
+
+        public bool IsDefault => _leafCharacters == null;
+        public bool IsEmpty => Length == 0;
+        public bool IsDefaultOrEmpty => IsDefault || IsEmpty;
+
+        public VirtualCharSequence GetSubSequence(TextSpan span)
+           => new VirtualCharSequence(
+               _leafCharacters, new TextSpan(_span.Start + span.Start, span.Length));
+
+        public VirtualChar First() => this[0];
+        public VirtualChar Last() => this[this.Length - 1];
+
+        public Enumerator GetEnumerator()
+            => new Enumerator(this);
+
+        public VirtualChar? FirstOrNullable(Func<VirtualChar, bool> predicate)
+        {
+            foreach (var ch in this)
+            {
+                if (predicate(ch))
+                {
+                    return ch;
+                }
+            }
+
+            return null;
+        }
+
+        public bool Contains(VirtualChar @char)
+            => IndexOf(@char) >= 0;
+
+        public int IndexOf(VirtualChar @char)
+        {
+            int index = 0;
+            foreach (var ch in this)
+            {
+                if (ch == @char)
+                {
+                    return index;
+                }
+
+                index++;
+            }
+
+            return -1;
+        }
+
+        public string CreateString()
+        {
+            var builder = PooledStringBuilder.GetInstance();
+            foreach (var ch in this)
+            {
+                builder.Builder.Append(ch.Char);
+            }
+
+            return builder.ToStringAndFree();
+        }
+
+        [Conditional("DEBUG")]
+        public void AssertAdjacentTo(VirtualCharSequence virtualChars)
+        {
+            Debug.Assert(_leafCharacters == virtualChars._leafCharacters);
+            Debug.Assert(_span.End == virtualChars._span.Start);
+        }
+
+        /// <summary>
+        /// Combines two <see cref="VirtualCharSequence"/>s, producing a final
+        /// sequence that points at the same underlying data, but spans from the 
+        /// start of <paramref name="chars1"/> to the end of <paramref name="chars2"/>.
+        /// </summary>  
+        public static VirtualCharSequence FromBounds(
```

Comment:
>FromBounds [](start = 42, length = 10)

Can/should this method assert that `chars1` and `chars2` are adjacent (with `AssertAdjacentTo`)?

Reply:
They're technically not.  You can merge N VirtualCharSequences in a row.  So the parser goes and finds the N of them (making sure each is adjacent to the last).  Then you do a FromBounds with the first/last one (which may not be touching).

Reply:
Alternatively (though i **really** don't care), since these are structs, we could just increase the size each time we get a new adjacent text node we want to grow.  BUt htat would be N (cheap) operations, instead of 1 cheap operation :)